neps/optimizers/models/deepGP.py

```diff
@@ -0,0 +1,572 @@
+        covar_x = self.covar_module(x)
+
+        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
+
+
+@dataclass
+class DeepGPDataTransformer:
+    # TODO: This class could be used for other models as well
+    space: SearchSpace
+    fidelity_bounds: tuple[int | float, int | float] | None
+    normalize_y: bool
+    min_learning_curve_length: int
+    learning_curve_pad_value: float
+    device: torch.device
+
+    numericals: dict[str, FloatParameter | IntegerParameter] = field(init=False)
+    categoricals: dict[str, CategoricalParameter] = field(init=False)
+    output_dim: int = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.numericals = {
+            name: h
+            for name, h in self.space.items()
+            if isinstance(h, (FloatParameter, IntegerParameter))
+        }
+        self.categoricals = {
+            name: h
+            for name, h in self.space.items()
+            if isinstance(h, CategoricalParameter)
+        }
+        self.output_dim = len(self.numericals) + sum(
+            len(c.choices) for c in self.categoricals.values()
+        )
+
+    def encode_configs(
+        self,
+        configs: list[SearchSpace],
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        x_buffer = torch.empty(
+            (len(configs), self.output_dim),
+            device=self.device,
+            dtype=torch.float32,
+        )
+
+        # Normals are just fill the columns with the normalized values
+        for i, (hp_name, hp) in enumerate(self.numericals.items()):
+            budget_tensor = torch.tensor(
+                [config[hp_name].value for config in configs],
+                device=self.device,
+                dtype=torch.float32,
+            )
+
+            x_buffer[:, i] = (budget_tensor - hp.lower) / (hp.upper - hp.lower)
+
+        # Categoricals is a bit harder, we create a tensor with all the indices (values)
+        # as we did above, but then we sub-select the portion of the buffer for that categorical
+        # before inserting the one-hot encoding.
+        offset = len(self.numericals)
+        for hp_name, hp in self.categoricals.items():
+            budget_tensor = torch.tensor(
+                [config[hp_name]._value_index for config in configs],  # type: ignore
+                device=self.device,
+                dtype=torch.float64,
+            )
+
+            # .. and insert one-hot encoding (ChatGPT solution, verified locally)
+            portion = x_buffer[:, offset : offset + len(hp.choices)]
+            portion.scatter_(1, budget_tensor.unsqueeze(1), 1)
+
+            offset += len(hp.choices)
+
+        # Finally, ... budgets
+        budgets = [config.fidelity.value for config in configs]  # type: ignore
+        budget_tensor = torch.tensor(budgets, device=self.device, dtype=torch.float32)
+        if self.fidelity_bounds:
+            assert self.space.fidelity is not None
+            _min = self.space.fidelity.lower
```

Comment:
We've changed the `_min` to be `0` for budget for the default case, [here](https://github.com/automl/neps/blob/d6e123faf41dde2461241a20b2014e4345d9a2a3/neps/optimizers/bayesian_optimization/models/deepGP.py#L317). This gave us some performance gains and is how the budgets were normalized in the original implementation.

Reply:
I feel like this could be problematic, as what if the fidelity bounds are on the scale of `100-200`? Generally the bounds should be `0` though I would imagine.

Perhaps some kind of std normalization would be more suitable but it's a bit out of scope of this PR. I will leave it at `fidelity.lower` just for generality and not tune it to the experiments from the paper.

Reply:
I agree that the bounds should be from `0` in general. For this special case, I think the authors wanted to avoid feeding many `0`s to the NN (especially early on the optimization minimum budget will dominate the training dataset). So they went with assuming `100` is the half of the max fidelity, which is true when the budget is the number of epochs but not true for general search-spaces. 

I think we can keep `fidelity.lower` till we find a validation method for deepGP performance